### DISCUSSION TITLE
JSON feed of packages

### DISCUSSION THREAD

**Discussion by arkoe** [tighten/novapackages#292] [Ideas]
Would there be any issues with having a json feed for all packages available, so it could be used for stuff like Discord bots to easily find a package? Similar to what Laravel has for their docs: https://laravel.com/docs/9.x/index.json

I wouldn't mind creating a PR for this feature.

**Comment by marcusmoore** [comment]
Hi @arkoe,

We do have an API available but it only returns 10 items at a time so I can see how that isn't ideal for your use case.

One concern is that returning all packages in one response could be slow and potentially rate-limited because the app is currently set up to fetch data from providers when a package is requested (the result is cached for a while but doing it for all packages at once would be costly).

But, if we slim down the data returned from this route we can get around that issue.

I threw this together to test out the response size and time and it's quick 😄 

```php
Route::get('packages.json', function () {
    return cache()->remember('all-packages-as-json', 1800, function () {
        return Package::all()->map(function ($package) {
            return [
                'name' => $package->display_name,
                'abstract' => $package->abstract,
                'url' => route('packages.show', [
                    'namespace' => $package->composer_vendor,
                    'name' => $package->composer_package,
                ]),
            ];
        });
    });
});
```

If I don't have time to clean it up (move to dedicated controller and test) and PR it today, feel free to PR yourself 😄

**Reply by marcusmoore** [reply]
@arkoe before I implement this, is this output what you were looking for?

```json
[
    {
        "name": "Tinker Tool",
        "abstract": "Use the power of Tinker within your Nova application.",
        "url": "http://novapackages.test/packages/beyondcode/nova-tinker-tool"
    },
    {
        "name": "Backup Tool",
        "abstract": "This Nova tool lets you list all backups, create a new backup, download a backup, delete a backup. It supports multiple filesystems and destinations.",
        "url": "http://novapackages.test/packages/spatie/nova-backup-tool"
    },
   ...more...
```

**Reply by arkoe** [reply]
@marcusmoore That looks absolutely amazing!

The attributes you've implemented seem fine for a basic integration I have in mind. Perhaps in the future something like `nova-version`, `tags` and/or `author` could be interesting. Up to you what you initially want to add, with what you've sent I'm already happy :).

**Reply by arkoe** [reply]
Created the PR: #301

**Comment by marcusmoore** [comment]
Done in #301
